Specify ability descriptions with propagationUpdater

mrhaandi opened the pull request:
Closes #6397, #6476.

Based on observations in https://github.com/vcmi/vcmi/issues/6476#issuecomment-3674375672 and what donners do anyway https://github.com/vcmi/vcmi/issues/6397#issuecomment-3670291635.

Looking at
https://github.com/vcmi/vcmi/blob/e2fd1cd04ac5fa525be365de0d5f08696778be46/lib/bonuses/CBonusSystemNode.cpp#L365-L367
the propagated bonus changes (and thus may require a different description) if there is a `propagationUpdater`.

Let's make two design decisions:
1. the description of a bonus with a `propagationUpdater` should be the description of the propagated bonus (not the propagator bonus).
Reasoning: when defining a bonus which influences others and needs an updater (say you create an artifact with Azure Dragon's fear inducing ability) you say what the final applied effect should be.
2. never show an **exported** bonus with `propagationUpdater` in the creature window.
Reasoning: the description can be wrong because "inducing an effect" is different from "being affected by an effect".

This would make bonus description mechanics precise.

Reason 1. necessitates that bonuses with original descriptions (not the actual propagated descriptions) are shown correctly, which can be done by a dummy bonus.

IvanSavenko commented:
Not much different from NONE bonus that was proposed in original issue.

And will break once unit has multiple abilities that need such descriptions

mrhaandi committed:
specify that a bonus with a propagationUpdater should have as description the applied bonus

- do not show exported bonuses with propagationUpdater in creature window (wrong descriptions)
- fix Azure Dragon bonus display
- instantiate metavariables in creature window

mrhaandi commented:
The main question, deciding whether this particular pull request is sensible, is:
- should the main description of an ability with a propagator describe (1) the propagator effect or (2) the propagated effect?

This PR suggests that the path of least resistance is describing (2) the propagated effect. This way, all current functionality can be implemented using existing methods, with custom types for descriptions of the propagator effect. No new mechanics needed.

If (1) the propagator effect is described by the main description, then we need some new mechanics (like description propagator updater) to display the propagated effect.

Until this question is settled, this PR should be a draft (and maybe not part of 1.7, if the timing is too close).

IvanSavenko commented:
Looking through draft PR's.

Any plans to finish this one?

Thinking about this - I still prefer approach using propagation updaters. From my point of view, the only question is how to show this to modder and to make it more modder-friendly?

What about this:

We add new Updater, DescriptionUpdater (for internal use, not exposed to mods directly), 
Instead bonus json config has field "propagatedDescription". If this field is present, string will be loaded just like regular description, however instead of storing it in bonus, game will generate DescriptionUpdater with this string and attach it to bonus, potentially appending any existing propagation updaters (via `addPropagationUpdater` call).

In theory, we can also make propagatedIcon in the same way, but not sure if there is any point in it.

This way, from modder point of view this feature is straightforward: bonus has propagated description. And any implementation details are hidden in engine. It also remove any need for hacks, like DESCRIPTION bonus.

...unless modders are OK with using custom bonuses for descriptions? In which case this PR can be closed

mrhaandi commented:
> Any plans to finish this one?

I think the current PR exists mainly for the purpose of discussion of one proposed workaround for the mentioned bugs for the time being, and has strong drawbacks. So I'm currently not in favor of pursuing this specific approach.

The field "propagatedDescription" could be cleaner, although I'm not sure how future proof it is, if bonuses interact. It feels like some part of internal bonus logic is still missing.
My current opinion: https://github.com/vcmi/vcmi/issues/6476#issuecomment-3674375672
Unfortunately, not too constructive :)

IvanSavenko commented:
> The field "propagatedDescription" could be cleaner, although I'm not sure how future proof it is, if bonuses interact. It feels like some part of internal bonus logic is still missing.

We might need different bonus accumulation logic.

In some cases it might make sense to accumulate bonuses. But in other cases - not really. Or replace accumulation with some grouping of bonuses, to show bonuses that affect the same thing together?

mrhaandi commented:
> We might need different bonus accumulation logic.

Maybe we can look at how others solve similar problems. For some inspiration:
- Unreal's Gameplay Ability System solves a similar problem and uses hierarchical Gameplay Tags and modifier operations.
- Magic The Gathering "layer system" has a comprehensive ruleset which identifies key concepts, exact mechanics (which some people still understand in depth), and specific wording.
